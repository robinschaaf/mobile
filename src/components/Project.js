import React, { Component } from 'react';
import {
  Alert,
  Animated,
  Image,
  Linking,
  Text,
  TouchableOpacity,
  View
} from 'react-native';
import EStyleSheet from 'react-native-extended-stylesheet'
import GoogleAnalytics from 'react-native-google-analytics-bridge'
import {Actions} from 'react-native-router-flux'
import Icon from 'react-native-vector-icons/FontAwesome'

class Project extends Component {
  constructor(props) {
    super(props)
    this.handleClick = this.handleClick.bind(this)
    this.state = {
      loading: true,
      fadeAnim: new Animated.Value(0)
    }
  }

  imageLoadEnd() {
    this.setState({ loading: false })
    Animated.timing(
      this.state.fadeAnim,
      { toValue: 1,
        duration: 200,
      },
    ).start();
  }


  handleClick() {
    Actions.Classify({ projectID: '3321', workflowID: '2725' })

    //GoogleAnalytics.trackEvent('view', this.props.project.display_name)

    //if (this.props.project.redirect) {
    //  this.openURL(this.props.project.redirect)
    //} else {
    //  Actions.ZooWebView({project: this.props.project})
    //}

  }

  openURL(url){
    Linking.canOpenURL(url).then(supported => {
      if (supported) {
        Linking.openURL(url);
      } else {
        Alert.alert(
          'Error',
          'Sorry, but it looks like you are unable to open the project in your default browser.',
        )
      }
    })
  }

  render() {
    const avatar =
      <Image source={{uri: `https://${this.props.project.avatar_src}`}} style={styles.avatar} onLoadEnd={ ()=>{ this.imageLoadEnd() } } />

    const defaultAvatar =
      <Image source={require('../../images/teal-wallpaper.png')} style={[styles.avatar, styles.defaultAvatar]} onLoadEnd={ ()=>{ this.imageLoadEnd() } } />

    return (
      <Animated.View style={{ opacity: this.state.fadeAnim }}>
        <TouchableOpacity
          activeOpacity={0.9}
          onPress={this.handleClick}
          style={styles.container}>
          { this.props.project.avatar_src ? avatar : defaultAvatar }
            <View style={styles.forBorderRadius} />
            <View style={styles.textContainer}>
              <View style={styles.titleContainer}>
                <Text style={styles.title} numberOfLines={1} ellipsizeMode={'tail'}>{this.props.project.display_name}</Text>
                <Text style={styles.description} numberOfLines={2} ellipsizeMode={'tail'}>{this.props.project.description}</Text>
              </View>
              <View style={[styles.iconContainer, { backgroundColor: this.props.color }]}>
                <Icon name="angle-right" style={styles.icon} />
              </View>
            </View>
        </TouchableOpacity>
      </Animated.View>
    )
  }
}

const styles = EStyleSheet.create({
  $boxHeight: 232,
  $titleHeight: 88,
  $borderRadius: 5,
  $iconSize: 40,
  $sidePadding: 15,
  $totalSidePadding: '$sidePadding * 2',
  $subtractTextWidth: '48 + $iconSize',
  container: {
    height: '$boxHeight + 12',
    marginHorizontal: 10,
    marginBottom: 25,
  },
  forBorderRadius: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
    height: '$titleHeight',
    backgroundColor: '$lightGreyBackground'
  },
  avatar: {
    borderRadius: '$borderRadius',
    borderWidth: 1,
    borderColor: '$greyBorder',
    flex: 1,
    height: '$boxHeight',
    resizeMode: 'cover'
  },
  defaultAvatar: {
    height: null,
    width: null,
  },
  textContainer: {
    position: 'absolute',
    bottom: 0,
    borderBottomLeftRadius: '$borderRadius',
    borderBottomRightRadius: '$borderRadius',
    borderWidth: 1,
    borderColor: '$greyBorder',
    borderTopColor: 'transparent',
    backgroundColor: 'white',
    flex: 1,
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingLeft: 15,
    paddingRight: 15,
    height: '$titleHeight',
  },
  titleContainer: {
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
    paddingRight: 20,
    width: '100% - $subtractTextWidth',
  },
  title: {
    backgroundColor: '$transparent',
    color: '$darkGreyTextColor',
    fontFamily: 'OpenSans-Semibold',
    fontSize: 18,
    paddingBottom: 3
  },
  description: {
    color: '$greyTextColor',
    fontFamily: 'OpenSans',
    fontSize: 12,
  },
  iconContainer: {
    borderRadius: '0.5 * $iconSize',
    alignItems: 'center',
    justifyContent: 'flex-start',
    width: '$iconSize',
    height: '$iconSize'
  },
  icon: {
    backgroundColor: 'transparent',
    color: 'white',
    fontSize: 36,
    lineHeight: 37,
    paddingLeft: 4
  }
});

Project.propTypes = {
  project: React.PropTypes.object.isRequired,
  color:  React.PropTypes.string.isRequired,
}
export default Project
