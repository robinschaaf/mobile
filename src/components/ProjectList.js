import React from 'react'
import {
  ListView,
  View
} from 'react-native'
import EStyleSheet from 'react-native-extended-stylesheet'
import StyledText from './StyledText'
import Project from './Project'
import NavBar from './NavBar'
import { connect } from 'react-redux'
import GoogleAnalytics from 'react-native-google-analytics-bridge'

GoogleAnalytics.trackEvent('view', 'Project')

const mapStateToProps = (state) => ({
  user: state.user,
  isConnected: state.isConnected,
  dataSource: dataSource.cloneWithRows(state.projectList[state.selectedProjectTag])
})

const dataSource = new ListView.DataSource({
  rowHasChanged: (r1, r2) => r1 !== r2,
})

class ProjectList extends React.Component {
  constructor(props) {
    super(props)
  }

  renderRow(project, color) {
    return (
      <Project project={project} color={color} />
    );
  }

  static renderNavigationBar() {
    return <NavBar title={'Projects'} showBack={true} />;
  }

  render() {
    const projectList =
      <ListView
        dataSource={this.props.dataSource}
        renderRow={(rowData) => this.renderRow(rowData, this.props.color)}
        enableEmptySections={true}
      />

    const noConnection =
      <View style={styles.messageContainer}>
        <StyledText textStyle={'errorMessage'}
          text={'You must have an internet connection to use Zooniverse Mobile'} />
      </View>

    return (
      <View style={styles.container}>
        <View style={styles.innerContainer}>
          { this.props.isConnected ? projectList : noConnection }
        </View>
      </View>
    );
  }
}

const styles = EStyleSheet.create({
  container: {
    flex: 1,
    paddingTop: 60,
    backgroundColor: '$lightGreyBackground'
  },
  innerContainer: {
    flex: 1,
    marginTop: 30
  },
  listStyle: {
    paddingTop: 90
  },
  messageContainer: {
    padding: 15,
  },
});

ProjectList.propTypes = {
  user: React.PropTypes.object,
  isConnected: React.PropTypes.bool,
  dataSource: React.PropTypes.object,
  tag: React.PropTypes.string,
  color: React.PropTypes.string,
  fetchProjects: React.PropTypes.func
}

export default connect(mapStateToProps)(ProjectList)
