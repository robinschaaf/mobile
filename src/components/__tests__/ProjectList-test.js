import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { ProjectList } from '../ProjectList'
import Project from '../Project'

const project = {
  avatar_src: 'fake_avatar',
  display_name: 'Nice project',
  workflows: [{swipe_verified: true}],
  tags: ['nature']
}

const selectedProjectTag = 'nature'
const projectList = [project]

jest.mock('ListView', () => require('react').createClass({
    statics: {
        DataSource: require.requireActual('ListView').DataSource,
    },
    render() {
        return <Project project={project} color={'#AFA48C'}/>
    },
}))

it('renders correctly', () => {
  const tree = renderer.create(
    <ProjectList projectList={projectList} selectedProjectTag={selectedProjectTag} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
