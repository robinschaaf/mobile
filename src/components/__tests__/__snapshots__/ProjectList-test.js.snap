exports[`test renders correctly 1`] = `
<View
  style={undefined}>
  <View
    style={undefined}>
    <View
      style={
        Object {
          "height": 269,
          "opacity": 0,
        }
      }>
      <TouchableOpacity
        activeOpacity={0.9}
        onPress={[Function]}
        style={
          Array [
            undefined,
            Object {
              "height": 244,
            },
          ]
        }>
        <Image
          onLoadEnd={[Function]}
          source={
            Object {
              "uri": "fake_avatar",
            }
          }
          style={
            Array [
              undefined,
              Object {
                "height": 244,
              },
            ]
          } />
        <View
          style={undefined} />
        <View
          style={undefined}>
          <View
            style={undefined}>
            <View
              style={undefined}>
              <View
                style={undefined}>
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  numberOfLines={1}
                  style={undefined}>
                  Nice project
                </Text>
              </View>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                numberOfLines={2}
                style={undefined} />
            </View>
            <View
              style={
                Array [
                  undefined,
                  Object {
                    "backgroundColor": "#AFA48C",
                  },
                ]
              }>
              <Text
                accessible={true}
                allowFontScaling={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontFamily": "FontAwesome",
                      "fontSize": 12,
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    undefined,
                  ]
                }>
                ï„…
              </Text>
            </View>
          </View>
        </View>
      </TouchableOpacity>
    </View>
  </View>
</View>
`;
